src/session.js

```diff
@@ -1,545 +1,182 @@
+    session.rpc.on('message', session.onMessage.bind(session));
+    session.rpc.on('notification', session.onNotification.bind(session));
+    session.rpc.on('traffic', session.onTraffic.bind(session));
+    session.on('handle-changed', handle => session.apis.onHandleChanged(handle));
+    session.on('handle-closed', handle => session.apis.onHandleClosed(handle));
+    session.on('closed', () => session.apis.onSessionClosed());
+    Object.keys(session.eventListeners || {})
+      .forEach(key => session.on(key, session.eventListeners[key]));
+    session.emit('session-created', session);
+  }
+
+  /**
+  * Event handler for re-triggering error events from RPC.
+  * @emits socket-error
+  * @param {Error} err Webocket error event.
+  */
+  onError(err) {
+    if (this.suspendResume.isSuspended) {
+      return;
+    }
+    this.emit('socket-error', err);
+  }
 
   /**
-  * Constructor
-  * @param {Object} rpc - The RPC instance used by the session.
-  * @param {Boolean} delta=true - Flag to determine delta handling.
-  * @param {Object} definition - The definition instance used by the session.
-  * @param {Object} JSONPatch - JSON patch object.
-  * @param {Function} Promise - The promise constructor.
-  * @param {Object} listeners - A key-value map of listeners.
-  * @param {Array} interceptors - An array of interceptors.
-  * @param {Boolean} suspendOnClose - when true, the session will be suspended if the underlying
-  *                                   websocket closes unexpectedly.
+  * Event handler for the RPC close event.
+  * @emits suspended
+  * @emits closed
+  * @param {Event} evt WebSocket close event.
   */
-  constructor(rpc, delta, definition, JSONPatch, Promise, listeners = {},
-    interceptors = [], suspendOnClose) {
-    Events.mixin(this);
-    this.rpc = rpc;
-    this.delta = delta;
-    this.definition = definition;
-    this.JSONPatch = JSONPatch;
-    this.Promise = Promise;
-    this.apis = new ApiCache();
-    this.suspendOnClose = suspendOnClose;
-    this.connectionId = connectionIdCounter += 1;
-    this.responseInterceptors = [{
-      onFulfilled: this.processErrorInterceptor,
-    }, {
-      onFulfilled: this.processDeltaInterceptor,
-    }, {
-      onFulfilled: this.processResultInterceptor,
-    }, {
-      onFulfilled: this.processMultipleOutParamInterceptor,
-    }, {
-      onFulfilled: this.processOutInterceptor,
-    }, {
-      onFulfilled: this.processObjectApiInterceptor,
-    }];
-    this.responseInterceptors.push(...interceptors);
-    this.registerRpcListeners();
-
-    this.suspended = false;
-
-    this.on('handle-changed', (handle) => {
-      const api = this.apis.getApi(handle);
-      if (api) {
-        api.emit('changed');
-      }
-    });
-
-    this.on('handle-closed', (handle) => {
-      const api = this.apis.getApi(handle);
-      if (api) {
-        api.emit('closed');
-        this.apis.remove(handle);
-      }
-    });
-
-    this.on('suspended', () => {
```

Comment:
By removing this, where do you suspend if the socket fails? before we emitted the suspended event on both manual and "surprise" suspensions. and the suspended flag was set here.